cs/src/core/RuntimeDeserializer.cs

```diff
@@ -0,0 +1,671 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Bond
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using Bond.IO.Safe;
+    using Bond.Protocols;
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a runtime schema.
+    /// </summary>
+    public static class RuntimeDeserialize
+    {
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from a tagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The tagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableTaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from an untagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The untagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableUntaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+    }
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a <see cref="RuntimeSchema"/>.
+    /// </summary>
+    public class RuntimeDeserializer
```

Comment:
This is my first time working with expression trees, and figuring out how everything works has been slow for me (not having source code to debug hurts!).

I agree that a more general solution like those you describe would be more appropriate (especially with there already being interest expressed in JSON and XML). With me not being very proficient with expressions and likely having higher-priority work to do during the day, I'm not sure how long it'll take me to change my implementation to use expressions (or maybe IProtocolWriter).

I opted not to cache deserializers on a per-schema basis primarily because SchemaDef objects are mutable.